### PR TITLE
New feature support for DataFrameConnector, NormalizedFrequencyEncoder & NormalizedLabelEncoder; CTGAN Optimization and Performance Enhancements.

### PR DESCRIPTION
**Pull request by cyantangerine**
## Description

1. Performance improvement: The performance improvement of Disk_cache.
For pd.concat, Iterative connections are slower than one-time connections because each connection requires calculating the index.
2. New feature: The DataFrameConnector can be used for smaller datasets, all loaded in memory, without the need to store on disk causing performance loss. It can reduce disk damage for large amounts of small data.
The DataFrameConnector is a proxy direct to connect DataFrame.
2.1 Simultaneously update the NoCache logic of DataLoader corresponding to DataFrameConnector, removed disk file solving time.
2.2 Simultaneously supports whether NDArrayLoader allows save_to_file.
3. New feature: NormalizedFrequencyEncoder and NormalizedLabelEncoder, supports encoding categorical data to a float number, can significantly reduce training dimensions, greatly improve training speed, and significantly reduce training memory for large datasets. At the same time, the data quality has been evaluated by SDV with minimal reduction.
For NormalizedLabelEncoder, in fit, using sorted unique values, it will create a [-1,1] key to value a map by uniform distribution. In transform, it maps the value to key. In reverse_transform, it find a nearest key to reverse-map the key to value.
3.1 Update Metadata simultaneously to support specifying encoders for table columns. Furthermore, it supports checking independent values using categorical_threshold and automatically selecting encoders for table columns.
The type of encoders are 'label' and 'onehot' now. If the column's unique count > threshold or it's encoder has been specified as 'label', the CTGAN model will choose the 'NormalizedLabelEncoder' to transform the column.  
3.2 Add a linear activation function in CTGAN to adapt to label encoding.
4. Code structure optimization: Provide a BatchedSynthesizerModel class to support models generated by Batched. Rename batch_size in TVAE to match CTGAN's _batch_size.
5. Generation performance optimization: CTGAN.sample supports setting the drop_more parameter to provide more generated content for the Synthesizer's sample at a time, reducing the loss caused by excessive generation. At the same time, in order to prevent the calculation of excessively generated values from being too small, it is considered to take the maximum value of batch_size with BatchedSynthesizer (description 4).
6. New feature: Loading models from disk allows for device specification instead of default CUDA if available.
By added **kwargs in load function to allow using model_kwargs (exist arg) in  model.__init__. Device param is included. 
7. Fix: numeric inspector error when using int32 or float32.
Instead checking type in array by using pandas type-checking api.
8. Performance: DatetimeFormatter using mean to instead of zero in timestamp.
If the formatter go error, it replace the value as zero in the past. Now, we replace it as  NaN, after all the value replaced, we fill the NaN and NULL as the mean of all values.
9. Performance: If we used DatetimeFormatter, remove the successful formatted datetime columns from metadata.discrete_columns and change it type to float.
## Motivation and Context

Description 2 solve the problem mentioned in #246 .
Description 3 partially solve the problem mentioned in #77 .
## How has this been tested?

Some tests are been given.

## Types of changes

- [ ] Maintenance (no change in code, maintain the project's CI, docs, etc.)
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: sdgx/data_connectors/dataframe_connector.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+import os
+from functools import cached_property
+from typing import Callable, Generator
+
+import pandas as pd
+
+from sdgx.data_connectors.base import DataConnector
+
+
+class DataFrameConnector(DataConnector):
```
Comment (jalr4ever) [inline sdgx/data_connectors/dataframe_connector.py:12]: I think we should add more code comments here.

File: sdgx/models/components/optimize/sdv_ctgan/data_transformer.py [hunk 2]
```diff
@@ -79,15 +84,30 @@ def _fit_discrete(self, data):
...
+        encoder.fit(data, column_name)
+        num_categories = len(encoder.dummies)
+        activate_fn = "softmax"
+
+        checked = (
+            self.metadata.check_categorical_threshold(num_categories) if self.metadata else False
+        )
+        if encoder_type == "onehot" or not checked:
+            pass
+        elif encoder_type == "label":
```
Comment (jalr4ever) [inline sdgx/models/components/optimize/sdv_ctgan/data_transformer.py:98]: Have you ever try Frequency Encoder? I have wrote implement in `StatisticDataTransformer`. Most of my case, it seems that better quality than label encoding.

File: sdgx/models/components/optimize/sdv_ctgan/data_transformer.py [hunk 3]
```diff
@@ -214,9 +238,12 @@ def inverse_transform(self, data, sigmas=None):
         Output uses the same type as input to the transform function.
         Either np array or pd dataframe.
         """
+
+        # TODO using pd.df.apply to increase performance.
```
Comment (jalr4ever) [inline sdgx/models/components/optimize/sdv_ctgan/data_transformer.py:242]: Remove if it's useless.

File: sdgx/synthesizer.py [hunk 4]
```diff
@@ -389,15 +393,26 @@ def _sample_once(
 
         """
         missing_count = count
-        max_trails = 50
+        max_trails = 5
```
Comment (jalr4ever) [inline sdgx/synthesizer.py:396]: Why it's change back to `5` and `missing_count * 1.2` ? We need more `max_trails`.

**Review by jalr4ever — CHANGES_REQUESTED** [review]
@cyantangerine Thanks for your contibution! It's a heavy PR. I think it's better to supply more info in the `PR Description`. This will also help people better understand the mechanisms and logic in the code!

`PR Description Reference` : https://github.com/hitsz-ids/synthetic-data-generator/pull/236 ; https://github.com/hitsz-ids/synthetic-data-generator/pull/219

**Comment by jalr4ever** [comment]
Relate Issue: https://github.com/hitsz-ids/synthetic-data-generator/issues/246

File: sdgx/models/components/optimize/sdv_ctgan/data_transformer.py (hunk 2 above)
Comment (cyantangerine) [reply sdgx/models/components/optimize/sdv_ctgan/data_transformer.py:98]: Thanks for recommendation. I will add a FrequencyEncoder choice for it.

File: sdgx/models/components/optimize/sdv_ctgan/data_transformer.py (hunk 3 above)
Comment (cyantangerine) [reply sdgx/models/components/optimize/sdv_ctgan/data_transformer.py:242]: That's a TODO for lastweek issue. Can I left it on code?

File: sdgx/synthesizer.py (hunk 4 above)
Comment (cyantangerine) [reply sdgx/synthesizer.py:396]: For small dataset, *4 may cause long time to generate especially in CTGAN? I have used the batch size in once sampling number to improve 'max_trails'. for other model that may need more trail?


Can i use this following?
max_trails = 50
batched model -> *1.2 for once sample
non_batched model -> *4 for once sample

**Comment by cyantangerine** [comment]
Relative issue: #77 for changing description 3.

File: sdgx/synthesizer.py (hunk 4 above)
Comment (jalr4ever) [reply sdgx/synthesizer.py:396]: > For small dataset, *4 may cause long time to generate especially in CTGAN? I have used the batch size in once sampling number to improve 'max_trails'. for other model that may need more trail?
> 
> Can i use this following? max_trails = 50 batched model -> *1.2 for once sample non_batched model -> *4 for once sample

**Background of the issue I hava**
Once we sampling from model, it's actually a post-process by `procrssors`, which like:
```python
            for d in self.data_processors:
                sample_data = d.reverse_convert(sample_data)
```

And I find that if we set max_trails to 5, the sampled data will be much smaller than the sample size specified by the user (In some cases of data).

**Question**
So if the `batch_size` is increased, will there still be enough sample size after processing and filtering by the processors? This is what I want to understand and confirm.

Comment (cyantangerine) [reply sdgx/synthesizer.py:396]: For why it's 1.2 not 4.
If we need 1000 sample, using 1.2 and batchsize=500, it will raw generated 1500 (cause we use 1200 to sample firstly), it still have 500 remain to filter. In some time, it enough.
But if we use 4, in that situation, it will take 2.5x extra times.

Comment (jalr4ever) [reply sdgx/synthesizer.py:396]: > Now let's talk the batchsize.
> 
> Let's think for the original design. When the missing count decreasing by cycle, only to multiply it is not a good plan, cause it converge to zero. Such as the missing=50, we use _4, model raw generated 200, then when reversed, it also has a probability still to missing. But in actually, the batched model can only generated by whole batch, if the size=500, 500-200=300 samples are ignored. If we select the max of missing_multiply and batchsize, it can reduce ignored sample, and reduce the missing probability after this trail.
> 
> So, i add 'drop more' argument in ctgan, then in sampling if the missing_multiply is not a multiple of batch size, it will be a ceil of it. And if missing_multiply < batchsize, it will be a batchsize.
> 
> So, for ctgan, we can also remove this code change. But for other model in batched, this method can do some favor, if the model not has drop_more (such as tvae, we can create an issue).

I see. To summarize, this issue arises from the lack of variable association between the `batch_size` of `Synthesizer.sample()` and `model.sample()`, resulting in the `batch_size` set by Synthesizer being smaller than the number of samples from `model.sample()`, causing excess samples from model.sample() to be wasted.

For this issue, I think the best solution is to refactor the relevant code so that this `batch_size` variable becomes a transferable context information, which can be perceived by both the model and the Synthesizer. Perhaps we could add a TODO.

Comment (cyantangerine) [reply sdgx/synthesizer.py:396]: Yes. But it is not all model are sampling in batch.

How about the Change Description 4?
(Created a subclass for batched model)

Comment (jalr4ever) [reply sdgx/synthesizer.py:396]: > Yes. But it is not all model are sampling in batch.
> 
> How about the Change Description 4? (Created a subclass for batched model)

I think it's good for now.

File: sdgx/models/components/sdv_rdt/transformers/categorical.py [hunk 5]
```diff
@@ -540,6 +542,79 @@ def _reverse_transform(self, data):
         return data.round().map(self.values_to_categories)
 
 
+class NormalizedLabelEncoder(LabelEncoder):
```
[the remark this answers is not in the document]
Comment (jalr4ever) [reply sdgx/models/components/sdv_rdt/transformers/categorical.py:545]: > The NormalizedFrequencyEncoder may need for test.
Yes it need. And  as a user, I am now actually confused about how to apply `LabelEncoding` or `FrequencyEncoding` in CTGAN. 

Here I suggest:

- Support test cases and assert its model encoder matched.
- Update use cases for changing model encoder with CTGAN

How to update the doc ? Reference PR: https://github.com/hitsz-ids/synthetic-data-generator/pull/236/files

FYI, when I updated the docs, I couldn't find the nice rst editor, I used Typora to write `md` and then exported it as `.rst`.

**Review by jalr4ever — CHANGES_REQUESTED** [review]

**Comment by Wh1isper** [comment]
Thank you for your hard work on this! This is a large Pull Request, and it would be much easier to review if it were split into several smaller PRs. Specifically, combining performance improvements with new features may lead us to spend more time discussing the implementation of the new features, which could delay the merging of the performance improvements.

I may find some time this weekend to review it!

File: sdgx/data_connectors/dataframe_connector.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+            length = self.df.shape[0]
+            if offset < length:
+                current = offset
+                while current < length:
+                    yield self.df.iloc[current : min(current + chunksize, length)]
+                    current += chunksize
+
+        return generator()
+
+    def reset_dataframe(self, df: pd.DataFrame):
```
Comment (Wh1isper) [inline sdgx/data_connectors/dataframe_connector.py:62]: Seems no reason to provide this feature, user can just renew a connector

File: sdgx/models/components/optimize/ndarray_loader.py [hunk 7]
```diff
@@ -71,8 +98,12 @@ def get_all(self) -> ndarray:
         return np.concatenate([array for array in self.iter()], axis=1)
 
     @cached_property
+    def __shape_0(self):
+        return self.load(0).shape[0]
+
+    @property
     def shape(self) -> tuple[int, int]:
-        return (self.load(0).shape[0], self.store_index)
+        return self.__shape_0, self.store_index
```
Comment (Wh1isper) [inline sdgx/models/components/optimize/ndarray_loader.py:106]: Seems like there's no need to do that? `__shape_0` is a little bit confusing as a property

File: sdgx/models/components/optimize/sdv_ctgan/data_transformer.py [hunk 8]
```diff
@@ -79,15 +86,57 @@ def _fit_discrete(self, data):
...
+            encoder.fit(data, column_name)
+            num_categories = 1
+            activate_fn = "liner"
+        elif encoder_type == "frequency":
+            encoder = NormalizedFrequencyEncoder()
+            encoder.fit(data, column_name)
+            num_categories = 1
+            activate_fn = "liner"
+        else:
+            raise ValueError(
```
Comment (Wh1isper) [inline sdgx/models/components/optimize/sdv_ctgan/data_transformer.py:128]: It's better we have a function here, like `fit_encoder`

File: sdgx/data_connectors/dataframe_connector.py (hunk 6 above)
Comment (cyantangerine) [reply sdgx/data_connectors/dataframe_connector.py:62]: When we use DataLoader to wrap it, such as we processing the data, it's make a lots of copy to recreat new Connector, Loader, etc. We can use this function apply the processed data directly. This function has been used in my MultiTableModel.

File: sdgx/models/components/optimize/ndarray_loader.py (hunk 7 above)
Comment (Wh1isper) [reply sdgx/models/components/optimize/ndarray_loader.py:106]: If so, `__shape_0` probably shouldn't be cached either?

File: sdgx/data_loader.py [hunk 9]
```diff
@@ -99,6 +100,9 @@ def __init__(
         cacher_kwargs: None | dict[str, Any] = None,
         identity: str | None = None,
     ) -> None:
+        if isinstance(data_connector, DataFrameConnector) and cacher is None:
```
[the remark this answers is not in the document]
Comment (cyantangerine) [reply sdgx/data_loader.py:103]: No. The goal of using NoCache, is to promote performance. Cause the whole Data has been loaded into memory in DataFrame, it's not need to use cache.

File: sdgx/data_connectors/dataframe_connector.py (hunk 6 above)
Comment (Wh1isper) [reply sdgx/data_connectors/dataframe_connector.py:62]: Create `DataConnector` will cause copy? The `df` should be a reference (mutable object). I don't think replacing the original dataframe with a new one is a reasonable `reset` operation.

File: sdgx/models/components/optimize/ndarray_loader.py (hunk 7 above)
Comment (cyantangerine) [reply sdgx/models/components/optimize/ndarray_loader.py:106]: Not this, it's the 'store_index' can't be cached. The shape0 can be cached safety because it's can't be modified after initial.

File: sdgx/data_loader.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (Wh1isper) [reply sdgx/data_loader.py:103]: And for this reason, using `self.DEFAULT_CACHER = NoCache` might be better. 

Further, if we can confirm that Connector+NoCache is the best choice in any case, then we should ignore any `Cacher` passed.

**Comment by cyantangerine** [comment]
All the questions behind has been solved. Thanks for reviewing. @jalr4ever @Wh1isper

**Review by Wh1isper — APPROVED** [review]
LGTM

**Review by jalr4ever — APPROVED** [review]
🚀 Approved!
